### PR TITLE
Replace current session entry when reloading

### LINKED ISSUE #13123: reloading a page should not create a new pipeline
If I'm not mistaken, reloading a page creates a new pipeline, and I don't think it should.
I'm not sure. But at least it should not create a new history entry when `reload()` is called.

Today it's possible to load a page, reload it, a do `iframe.goBack()`.

So either re-use the pipeline or drop the old pipeline.

### PR DESCRIPTION
**Pull request by cbrewster**
This PR adds a replacement option when navigating. It replaces the current session history entry after a new page has been loaded. This will prevent reloading from adding a new entry to the session history.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #13123 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13167)

### CODE REVIEW

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -1930,10 +1936,19 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             // Add new pipeline to navigation frame, and return frames evicted from history.
             self.pipelines.get_mut(&frame_change.new_pipeline_id)
                           .map(|pipeline| pipeline.frame = Some(frame_id));
-            self.frames.get_mut(&frame_id).map(|frame| frame.load(frame_change.new_pipeline_id));
+            if frame_change.replace {
+                let entry = self.frames.get_mut(&frame_id).map(|frame| {
+                    frame.replace_current(frame_change.new_pipeline_id)
+                });
+                if let Some(entry) = entry {
```
Comment (paulrouget) [inline components/constellation/constellation.rs]: I'm not sure, but is entry ever supposed to be none? `unwrap`?

**Comment by paulrouget** [comment]
This looks good to me, and pass my local tests.

File: components/constellation/constellation.rs (hunk 1 above)
Comment (ConnorGBrewster) [inline components/constellation/constellation.rs]: I should probably warn, the `None` case will only happen if the frame lookup fails.

Comment (ConnorGBrewster) [inline components/constellation/constellation.rs]: Fixed.

**Comment by asajeffrey** [comment]
Nitpicking about adding some comments, apart from that lgtm! You don't have to add the comments, but I think they would help. You can r=me.

---

Reviewed 2 of 10 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/constellation/constellation.rs, line 1389 at r2](https://reviewable.io:443/reviews/servo/servo/13167#-KS2mcFQo3Pgh1qIxseV:-KS2mcFR_sF3NeAfaEy7:bi03ayh) ([raw file](https://github.com/servo/servo/blob/0520ee4b5af6fe504c0ac84780a5132a50fd50af/components/constellation/constellation.rs#L1389)):*
> ```Rust
>     }
> 
>     fn load_url(&mut self, source_id: PipelineId, load_data: LoadData, replace: bool) -> Option<PipelineId> {
> ```

A comment somewhere in this file about what `replace` is for would be nice. Perhaps on this function?

---

*[components/script/script_thread.rs, line 232 at r2](https://reviewable.io:443/reviews/servo/servo/13167#-KS2ndsMgk_4xbwch3l7:-KS2ndsMgk_4xbwch3l8:b-pxb3o7) ([raw file](https://github.com/servo/servo/blob/0520ee4b5af6fe504c0ac84780a5132a50fd50af/components/script/script_thread.rs#L232)):*
> ```Rust
>     /// Begins a content-initiated load on the specified pipeline (only
>     /// dispatched to ScriptThread).
>     Navigate(PipelineId, LoadData, bool),
> ```

A comment saying what the `bool` is for would be nice.

---

*[components/script_traits/lib.rs, line 462 at r2](https://reviewable.io:443/reviews/servo/servo/13167#-KS2nveSWoPMqU58inxW:-KS2nveSWoPMqU58inxX:b85sx09) ([raw file](https://github.com/servo/servo/blob/0520ee4b5af6fe504c0ac84780a5132a50fd50af/components/script_traits/lib.rs#L462)):*
> ```Rust
>     /// Whether this iframe is a mozbrowser iframe
>     pub frame_type: FrameType,
>     /// Wether this load should replace the current entry (reload)
> ```

A comment about `reload` not adding an entry to the session history would be nice.

---

*[components/script_traits/script_msg.rs, line 86 at r2](https://reviewable.io:443/reviews/servo/servo/13167#-KS2o7ZeReaxFGgR32x2:-KS2o7Zfxrta2ynb4a7a:b85sx09) ([raw file](https://github.com/servo/servo/blob/0520ee4b5af6fe504c0ac84780a5132a50fd50af/components/script_traits/script_msg.rs#L86)):*
> ```Rust
>     /// has been dispatched.
>     LoadComplete(PipelineId),
>     /// A new load has been requested, with an option to replace the current entry once loaded.
> ```

A comment about `reload` not adding an entry to the session history would be nice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13167#-:-KS2oKw5zemswgfq8ZtK:b-htiv1j)*

**Pull request merged**